internal/auth/providers/groups_cache.go

```diff
@@ -0,0 +1,59 @@
+package providers
+
+import (
+	"reflect"
+	"time"
+
+	"github.com/buzzfeed/sso/internal/pkg/groups"
+	"github.com/datadog/datadog-go/statsd"
+)
+
+type Cache interface {
+	Get(keys string)
+	Set(entries []string)
+	Purge(keys []string)
+}
+
+type GroupCache struct {
+	metrics  *statsd.Client
+	provider Provider
+	cache    *groups.LocalCache
+}
+
+func NewLocalCache(provider Provider, ttl time.Duration, statsdClient *statsd.Client, tags []string) *GroupCache {
+	return &GroupCache{
+		metrics:  statsdClient,
+		provider: provider,
+		cache:    groups.NewLocalCache(ttl, statsdClient, tags),
+	}
+}
+
+// ValidateGroupMembership wraps the provider's ValidateGroupMembership around calls to check local cache for group membership information.
+func (p *GroupCache) ValidateGroupMembership(email string, allowedGroups []string, accessToken string) ([]string, error) {
+	groupMembership, err := p.cache.Get(email)
+	if reflect.DeepEqual(groupMembership.UserGroupData.AllowedGroups, allowedGroups) && len(groupMembership.UserGroupData.MatchedGroups) > 0 {
```

Comment:
This is ALOT of logic on this one line, can we split this to multiplel lines? The `reflect.DeepEqual` will only return true if each entry in the structs are equal so we will want to sort the output before we compare them. We can store the output sorted in the cache and just sort the passed groups or sort both on demand.

Reply:
Ah, yes! Sorting the groups is something I had intended to add when looking back through this but clearly neglected to do so...so thank you for pointing this out!